### ISSUE TITLE
Connection upgrades (raw -> tls)

### ISSUE THREAD

**Issue by Teifion** [springlobby/springlobby#991]
Question 1: Why does SpringLobby start with raw TCP and then upgrade to TLS? Why not either stay as raw or start as TLS?

Question 2: Is there a way to get SL to not need to upgrade the connection?

Thanks!

**Comment by abma** [comment]
1.: https://github.com/spring/uberserver/pull/271
2. compile without ssl support: https://github.com/springlobby/springlobby/blob/master/src/tasserver.cpp#L543 

IMHO "raw" connections are deprecated, thats why it always "upgrades" to TLS.

**Comment by Teifion** [comment]
Cheers for the reply. My implementation of Uberserver is having issues with the upgrading the connection and as such I've been trying to find workarounds.

Would I be correct in reading this as there is no way to just start it in TLS mode and skip the upgrade process entirely?

**Comment by abma** [comment]
> Would I be correct in reading this as there is no way to just start it in TLS mode and skip the upgrade process entirely?

yes, atm it not possible as it requires changes on the server and on the client.

**Comment by abma** [comment]
> Cheers for the reply. My implementation of Uberserver is having issues with the upgrading the connection and as such I've been trying to find workarounds.

what issue do you have?

**Comment by abma** [comment]
a debug build of springlobby and running it with "springlobby -cl -l9" maybe helps to debug the problem?

**Comment by abma** [comment]
or https://github.com/spring/uberserver/blob/master/tests/tlsclient.py

**Comment by Teifion** [comment]
It won't upgrade the raw connection to a TLS one. I'm 99% certain the issue is on my end since the TCP server I'm using gives an error. Unfortunately SpringLobby won't run properly on my computer so I can't even test it locally.

I'm still digging into the issue but I've learned that looking at other avenues is a good idea, I started by asking on the [github issues of the TCP server I'm using.](https://github.com/ninenines/ranch/issues/315).

Chobby doesn't make use of the upgrade so it's not been an issue prior to now. Starting off with a TLS connection works fine btw, it's purely the upgrade part!

**Comment by Teifion** [comment]
Thank you for the help. I managed to get spring lobby working locally (Had to run "env" instead of "springlobby" version). I have fixed the TLS issue and it appears to be working.

The cause of the issue was 100% on my end. I was not updating a state variable with the new socket that'd been upgraded, meaning I was trying to send/listen on the old TCP socket.

**Comment by abma** [comment]
glad that i could help. i guess this change: https://github.com/Teifion/teiserver/commit/4083dd17b4719fff5848ef938fe627135a92616d#diff-9302d2bb7f6a55e1be9c26413e7b4a36fa8d377e55138cc3c96c6c0f5e4b13b3R111

last question before i close this issue: why did you write a lobby server from scratch, what is/was wrong with uberserver?

**Comment by Teifion** [comment]
Yep, that's the change!

I thought that it would be easier to rewrite it than maintain it. Elixir is designed for this sort of application so it was a sensible rewrite. It's taken about the amount of time expected but it'll be significantly faster to add some of the features I'm planning.

While I have you though, I plan to add a bunch of features (starting with matchmaking). What's the best and most helpful way to let you know how they're being implemented?

**Comment by silentwings** [comment]
I don't think we (i.e. spring infra maintainers) need to know what's going on in diverging infra of particular games, as long as the various licenses of any involved code is kept to.

A suggestion for you: if it was me writing a lobbyserver from scratch, I would think carefully about what I needed from it and then probably write a new protocol with more structured messages that more closely bound together (auto)hosts and server itself.

Sidenote: currently about 80% of the connections to lobby.springrts.com use TLS, so it seems the various subprojects connected to it so we are not yet in a position where we could enforce it - although imo it would be best to do so when it is possible.
2021-04-01 18:13:07 INFO  DataHandler.stats:718  Number of logins: 48899
2021-04-01 18:13:07 INFO  DataHandler.stats:719  TLS logins: 39007

**Comment by Teifion** [comment]
I was thinking more that if features are added to Teiserver beyond Uberserver it might mean people would be unhappy if SL didn't have the ability to communicate with those features. Given it's a community I wanted to ensure if I can make it easier for SL to keep up to date with new features I would.

Uberserver is indeed stable and at every juncture when discussing my implementation I'm taking pains to point this out. Uberserver is working for many people and I don't expect that to change any time soon.

I already have plans to put forwards a new lobby protocol when the time is appropriate but my expectation is the amount of work required on all sides will mean it would need to be run in parallel where possible. You're spot on with the O(n^2) problem, I did some load tests and managing that problem will be the big thing.

Thank you for the numbers on TLS, that's really interesting to know. If I had a new protocol it'd have enforcement of TLS as one of the steps!

Where does discussion on the SL and protocol stuff take place?

**Comment by silentwings** [comment]
The protocol is recorded fairly carefully in words in https://github.com/spring/LobbyProtocol/ but that repo is generally not used for discussing it. Talk on changes/extensions/etc gets drafted on the issue tracker of https://github.com/spring/uberserver and then made via PRs to that repo and to this one, which together defines exactly what the live protocol version is. 

I think its still fair to say that both these repos (+ most of the other infra they interface with) are maintained, in the sense that we'd fix critical bugs if they were found, but that is the sum total of activity.

**Comment by Teifion** [comment]
Awesome, the first new feature I'm adding will be matchmaking so I'll share my protocol proposal on the uberserver tracker and we'll see how we go! Thanks.

**Comment by PicassoCT** [comment]
Cant use springlobby to host a game anymore.
Upgraded to built from repo/master today. Still can not host.
 ** Server ** A TLS connection is required to host battles. Please upgrade your client.

**Comment by PicassoCT** [comment]
@abma What can i do to fix?
